syntax = "proto3";

package meshtrade.compliance.client.v1;

import "buf/validate/validate.proto";
import "google/type/date.proto";
import "meshtrade/compliance/client/v1/identification_document_type.proto";
import "meshtrade/compliance/client/v1/pep_status.proto";
import "meshtrade/compliance/client/v1/source_of_income_and_wealth.proto";
import "meshtrade/compliance/client/v1/tax_residency.proto";
import "meshtrade/type/v1/address.proto";
import "meshtrade/type/v1/contact_details.proto";

option go_package = "github.com/meshtrade/api/go/compliance/client/v1;client_v1";
option java_package = "co.meshtrade.api.compliance.client.v1";

/*
   NaturalPerson is the identity of an individual person.
   It contains the core, verifiable components of an individual's identity. which are
   verified against their passport, utility bills, government records etc. during Know Your Client (KYC) checks.
   Note on Required Fields: Fields marked as 'Required' are essential
   for a successful compliance check, but are not mandatory for creation.
*/
message NaturalPerson {
  /*
     The client's full legal name, as it appears on their official identification documents.
     Required for verification.
  */
  string full_name = 1 [(buf.validate.field) = {
    string: {
      max_len: 255
    }
  }];

  /*
     The client's date of birth.
     Required for verification.
  */
  google.type.Date date_of_birth = 2;

  /*
     The ISO 3166-1 alpha-2 country code of the client's nationality/citizenship.
     This is the two-letter country code (e.g., "ZA" for South Africa,
     "NL" for the Netherlands). The value should be in uppercase.
     See https://www.iso.org/iso-3166-country-codes.html for a full list.
     Required for verification.
  */
  string country_of_citizenship = 3 [(buf.validate.field) = {
    cel: {
      id: "country_code.format.optional",
      message: "country_of_citizenship must be empty or a valid ISO 3166-1 alpha-2 country code",
      expression: "size(this) == 0 || this.matches('^[A-Z]{2}$')"
    }
  }];

  /*
     An identification number for the client, as found on the provided document.
     The format is dependent on the identification_document_type.
     Required for verification.
  */
  string identification_number = 4 [(buf.validate.field) = {
    string: {
      max_len: 100
    }
  }];

  /*
     The type of identification document provided to prove the correctness of the
     given identification_number (e.g., Passport, Driver's License).
     Required for verification.
  */
  meshtrade.compliance.client.v1.IdentificationDocumentType identification_document_type = 5 [(buf.validate.field) = {
    enum: {defined_only: true}
  }];

  /*
     The expiration date of the identification document, if applicable.
     Required for verification if the document has an expiry date.
  */
  google.type.Date identification_document_expiry_date = 6;

  /*
     The client's primary physical residential address.
     If `postal_address` is not provided, this address will also be used for postal mail.
     Required for verification.
  */
  meshtrade.type.v1.Address physical_address = 7;

  /*
     The client's postal address, if it is different from the physical address.
     Optional for verification.
  */
  meshtrade.type.v1.Address postal_address = 8;

  /*
     The client's status as a Politically Exposed Person (PEP).
     This is a mandatory check for regulatory compliance.
     Required for verification.
  */
  meshtrade.compliance.client.v1.PepStatus pep_status = 9 [(buf.validate.field) = {
    enum: {defined_only: true}
  }];

  /*
     The individual's personal contact details (personal email, personal mobile).
     Optional for verification.
  */
  meshtrade.type.v1.ContactDetails contact_details = 10;

  /*
     The primary sources of the client's regular income (e.g., employment, pension).
     Required for verification.
  */
  repeated meshtrade.compliance.client.v1.SourceOfIncomeAndWealth sources_of_income = 11 [(buf.validate.field) = {
    repeated: {
      items: {
        enum: {defined_only: true}
      }
    }
  }];

  /*
     The origins of the client's total net worth or assets (e.g., inheritance, investments).
     This is distinct from the source of income.
     Required for verification.
  */
  repeated meshtrade.compliance.client.v1.SourceOfIncomeAndWealth sources_of_wealth = 12 [(buf.validate.field) = {
    repeated: {
      items: {
        enum: {defined_only: true}
      }
    }
  }];

  /*
     The client's tax residency information, required for CRS/FATCA reporting.
     A client can be a tax resident in multiple jurisdictions.
     Required for verification.
  */
  repeated meshtrade.compliance.client.v1.TaxResidency tax_residencies = 13;
}
